WeakAuras.lua

```diff
@@ -3138,9 +2980,268 @@ function WeakAuras.RegisterTriggerSystem(types, triggerSystem)
+  region.state = state;
+  if(region.SetDurationInfo) then
+    if (state.progressType == "timed") then
+      local now = GetTime();
+      local value = math.huge - now;
+      if (state.expirationTime and state.expirationTime > 0) then
+        value = state.expirationTime - now;
+      end
+      local total = state.duration or 0
+      local func = nil;
+
+      region:SetDurationInfo(total,
+                             now + value,
+                             func,
+                             state.inverseDirection);
+    elseif (state.progressType == "static") then
+      local value = state.value or 0;
+      local total = state.total or 0;
+      local durationFunc = state.durationFunc or true;
+
+      region:SetDurationInfo(value, total, durationFunc or true, state.inverseDirection);
+    else
+      -- Do nothing, should ideally clear duration info on region
+    end
+  end
+  if (state.resort) then
+    WeakAuras.ControlChildren(region.id);
+    state.resort = false;
+  end
+  if(region.SetName) then
+    region:SetName(state.name);
+  end
+  if(region.SetIcon) then
+    region:SetIcon(state.icon or "Interface\\Icons\\INV_Misc_QuestionMark");
+  end
+  if(region.SetStacks) then
+    region:SetStacks(state.stacks);
+  end
+  if(region.UpdateCustomText and not WeakAuras.IsRegisteredForCustomTextUpdates(region)) then
+    region.UpdateCustomText();
+  end
+  WeakAuras.UpdateMouseoverTooltip(region);
+  region:Expand();
+end
+
+-- Fallbacks if the states are empty
+local emptyTable = {}
+local emptyState = {};
+emptyState[""] = emptyTable;
+
+local function applyToTriggerStateTriggers(stateShown, id, triggernum)
+  if (stateShown and not triggerState[id].triggers[triggernum + 1]) then
+    triggerState[id].triggers[triggernum + 1] = true;
+    triggerState[id].triggerCount = triggerState[id].triggerCount + 1;
+    return true;
+  elseif (not stateShown and triggerState[id].triggers[triggernum + 1]) then
+    -- Check if any other clone is shown
+    local anyCloneShown = false;
+    for _, state in pairs(triggerState[id][triggernum]) do
+      if (state.show) then
+        anyCloneShown = true;
+        break;
+      end
+    end
+    if (not anyCloneShown) then
+      triggerState[id].triggers[triggernum + 1] = false;
+      triggerState[id].triggerCount = triggerState[id].triggerCount - 1;
+      return true;
+    end
+  end
+  return false;
+end
+
+local function evaluateTriggerStateTriggers(id)
+  if(triggerState[id].disjunctive == "any" and triggerState[id].triggerCount > 0
+      or (triggerState[id].disjunctive == "all" and triggerState[id].triggerCount == triggerState[id].numTriggers)
+      or (triggerState[id].disjunctive == "custom"
+          and triggerState[id].triggerLogicFunc
+          and triggerState[id].triggerLogicFunc(triggerState[id].triggers))
+    ) then
+    return true;
+  end
+  return false;
+end
+
+local toRemove = {};
+function WeakAuras.UpdatedTriggerState(id)
```

Comment:
This function is called by the BuffTrigger and GenericTrigger after they changed the triggerStates. It takes care of calling the above functions and showing/hiding regions as needed.